### ISSUE TITLE
Assignability for MarkRequired after upgrade from 9.1.2 to 9.3.0

### ISSUE THREAD

**Issue by akwodkiewicz** [ts-essentials/ts-essentials#349]
Similarly to #323, my type guard is not working correctly after the change inside `MarkRequired` (`Exclude` => `Omit`). What's more is that the addition of union handling (done in 9.3.0) has not fixed the problem. Here's the type guard with the error.

```ts
export function checkSetProperties<T, PropName extends keyof T>(
    obj: T,
    propNames: PropName[],
): obj is MarkRequired<T, PropName> {
//        ^^^^^^^^^^^^^^^^^^^^^^^^^
//        A type predicate's type must be assignable to its parameter's type.
//          Type 'MarkRequired<T, PropName>' is not assignable to type 'T'.
//            'T' could be instantiated with an arbitrary type which could be unrelated to 'MarkRequired<T, PropName>'.ts(2677)
    return !propNames.some(prop => !obj[prop]);
}
```

Link to TS Playground: [here](https://www.typescriptlang.org/play?#code/PQKhAIE4DoGZoAzhMAsAKAC4E8AOBTcAWQEMAnAawCV8BHAVwEsz8ATAHgBUAacKgaXD4AHpnwA7VgGdwFfNgD2AM3CcAfOAC8qoaInSdAfnAB5ALaNMXXgI0AyPnSYsOABUYBjCtb781GgC5wcXwAN3wyAG4MUAgYAEZoACZkNCw8QlJKGgZmNhMAGw4eX10xSRk5RRV1LXAAUWEPAvpWfB9bcAcc5zZ2dy8OvzVo9AwMJXpxD0xGBXFwDwALfC8AZXxMVzIFAjJZ-Ckfbd2AORIzQhFygyrlVTUACgxwV-AFACMAKyCeF7fcDtcOdLlIgidgRd8ABtAC63AwAEogp8vuBGDIstQnHlirwISD8BoAN7-V4sTD0MgLACEgLOUKk0CkCkuj3puC0GhpqOhHNhiNGAF9xuhJtNZvNFit1psIREDlJCnjwASoWV9JV5Pd1M90G93t9fgj9QCgYSwarzVC4SbkYa0RjiORsbkXMrjtbLiSyeAKVTaRyLczWfh2UCueAed8+UCBcLRcBgOAALRptPgACqawAggBxeqp9Pp0WMcRiMhKEgeQhrUMASXLESrNfApNNJCC4noZg+EVGrw+hi7Pb7UQwIvQHnmUkwDviQTrl0bFZbhG0xM74HiQvAGGn4lnDqSi4bTcr1fXba38V4HyCSUnGEYKkey1WFA2WyBCsYh2Vjyore4DQgA5B8oECoibb-EB0AfOASbBKOEQTkIBRSIQ7avHBCFId2vYROAAA+4BTG0ShlmwE6ikAA)

**Comment by Beraliv** [comment]
Hey @akwodkiewicz!

Thank you for reporting it!

It looks like a bug, let me take a closer look and I will be able to understand the fix

Meanwhile, you can use `T & MarkRequired<T, PropName>` instead `MarkRequired<T, PropName>` as a workaround:

```ts
function checkSetProperties<T, PropName extends keyof T>(
    obj: T,
    propNames: PropName[],
): obj is T & MarkRequired<T, PropName> {
    return !propNames.some(prop => !obj[prop]);
}
```

**Comment by akwodkiewicz** [comment]
Thank **you** for the quick reply and the workaround!

I'll be happy to read what is the reason for the faulty behaviour after you investigate, because I tried to understand what goes wrong and could not work it out myself 😅

**Comment by Beraliv** [comment]
I think this is related to this change: https://github.com/ts-essentials/ts-essentials/pull/324

Union support was added for all Mark-types therefore assignability was removed and that's why you need to add `T &` to make it assignable again

I will try to address it in the PR because I still think it's important to have assignability in place

**Comment by Beraliv** [comment]
Hey @akwodkiewicz!

I've created a PR with a fix, let me know if it starts working for you again, I plan to test it thoroughly once again and will merge it later next week

Let me know if you have any feedback

**Comment by akwodkiewicz** [comment]
It works for me 😄 Thank you for the PR!

**Comment by Beraliv** [comment]
Hey @akwodkiewicz, apologies for not looking at this for a long time, I will return to this ticket in the next weeks and I will keep you updated about my progress

**Comment by Beraliv** [comment]
I'm looking at this bug again and I'm going to think what's the preferable solution at the moment, I will keep you updated

**Comment by Beraliv** [comment]
Hey @akwodkiewicz! I'd like to fix it in v10 to prevent further regressions. I will look into it the first thing tomorrow and will hopefully merge it by the end of April

**Comment by Beraliv** [comment]
I merged the PR https://github.com/ts-essentials/ts-essentials/pull/364 but this required to remove support of TS@<4.5. The changes are included in the release and will be deployed next week - https://github.com/ts-essentials/ts-essentials/pull/386

**Comment by Beraliv** [comment]
It's released just now as a part of [v10.0.0](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.0)
